#Aim of this Project
This project aims to estimate the prices of vehicles with specific features at an automobile company based on existing car prices.

#Why Regression Problem
Regression is a key method used in **forecasting**. When it comes to making future predictions, using regression is always the better approach.

#Tensorflow Topics that are used.
* **TensorFlow Basics & Tensors:** Understanding multi-dimensional data structures for numerical computation.
Sequential Model API: Building a linear stack of layers for an end-to-end deep learning pipeline.
* **Dense (Fully Connected) Layers:** Implementing hidden layers to learn complex relationships between vehicle features.
* **Regression-Specific Architecture:** Designing an output layer without activation functions to predict continuous numerical values (Price).
* **Loss Functions (MSE):** Using Mean Squared Error to penalize large prediction errors during the learning process.
* **Optimizers (Adam):** Implementing adaptive moment estimation for efficient weight updates and faster convergence.
* **Evaluation Metrics (MAE):** Utilizing Mean Absolute Error to interpret the average error in actual currency (e.g., TL/USD).
* **Feature Scaling & Normalization:** Preparing data to ensure different scales (e.g., Mileage vs. Horsepower) do not bias the model.
* **Inference & Prediction:** Generating real-time price estimates for previously unseen car data.





In [ ]:
import numpy as np
import pandas as pd
from datetime import datetime
from sklearn.preprocessing import MinMaxScaler

current_year = datetime.now().year

np.random.seed(42)
n_vehicles = 1000

model_year = np.random.randint(2005,2024,n_vehicles)
mileage = np.random.randint(0,300000,n_vehicles)
engine_displacement = np.random.uniform(1.0,4.5,n_vehicles)
horsepower = np.random.randint(50,600,n_vehicles)
vehicle_age = current_year - model_year

x = np.column_stack((model_year,mileage,engine_displacement,horsepower,vehicle_age)).astype('float32')
y = (model_year*5000)+(horsepower*1000) - (mileage*10)
noise = np.random.normal(0,5000,n_vehicles)
y = (y + noise).astype('float32').reshape(-1,1)

x_scaled = x / x.max(axis=0)
y_scaled = y / y.max()


In [ ]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x_scaled,y_scaled,test_size=0.20, random_state=42)
print(f"x_train shape: {x_train.shape}")
print(f"y_train shape: {y_train.shape}")

x_train shape: (800, 5)
y_train shape: (800, 1)


In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

model = Sequential([
    Dense(64, activation='relu', input_shape=(5,)),
    Dense(32, activation='relu'),
    Dense(1)
])
model.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_3 (Dense)                 │ (None, 64)             │           384 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 32)             │         2,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 1)              │            33 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 2,497 (9.75 KB)

 Trainable params: 2,497 (9.75 KB)

 Non-trainable params: 0 (0.00 B)

In [ ]:
model.compile(optimizer='adam', loss='mse', metrics=['mae'])

In [ ]:
model.fit(x_train,y_train,epochs=100,batch_size=32, validation_data=(x_test,y_test), verbose=1)

Epoch 1/100
25/25 ━━━━━━━━━━━━━━━━━━━━ 2s 13ms/step - loss: 0.1151 - mae: 0.2755 - val_loss: 0.0297 - val_mae: 0.1381
Epoch 2/100
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - loss: 0.0146 - mae: 0.0977 - val_loss: 0.0075 - val_mae: 0.0731
Epoch 3/100
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - loss: 0.0049 - mae: 0.0576 - val_loss: 0.0023 - val_mae: 0.0395
Epoch 4/100
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - loss: 0.0015 - mae: 0.0305 - val_loss: 7.0270e-04 - val_mae: 0.0203
Epoch 5/100
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - loss: 4.1324e-04 - mae: 0.0158 - val_loss: 3.3321e-04 - val_mae: 0.0134
Epoch 6/100
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - loss: 2.4213e-04 - mae: 0.0119 - val_loss: 2.6199e-04 - val_mae: 0.0122
Epoch 7/100
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - loss: 1.9193e-04 - mae: 0.0106 - val_loss: 2.0335e-04 - val_mae: 0.0103
Epoch 8/100
25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - loss: 1.5230e-04 - mae: 0.0093 - val_loss: 1.6695e-04 - val_mae: 0.0090
Epoch 9/100
25/25 ━━━━━━━━━

In [7]:
test_loss, test_mae = model.evaluate(x_test,y_test,verbose=0)
print(f"Test Loss: {test_loss:.4f}")
print(f"Test MAE: {test_mae:.4f}")
#There is no overfitting since mae, val_loss and val_mae values look like accurate.

Test Loss: 0.0000
Test MAE: 0.0014


In [14]:
new_car = np.array([[2022,20000,3.2,300,current_year-2022]])
prediction =model.predict(new_car)
print(f"---New Vehicle Prediction---")
print(f"Inputs: {new_car}")
print(f"Prediction: {prediction}")

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 88ms/step
---New Vehicle Prediction---
Inputs: [[2.022e+03 2.000e+04 3.200e+00 3.000e+02 4.000e+00]]
Prediction: [[838.1323]]


#Model Performance Analysis and Insights
**Influential Features:** Vehicle price is typically most strongly correlated with Model Year and Mileage. The model's ability to capture this linear relationship can be directly interpreted through the MAE (Mean Absolute Error) value.

**Prediction Capacity:** The model tends to perform better at predicting "average" vehicle types (e.g., cars between 5-10 years old with standard mileage) that represent the bulk of the dataset.

**Limitations:** The model’s margin of error will likely increase for luxury vehicles or cars with extreme variables not included in the dataset, such as accident history or specific trim levels. This is because the model focuses exclusively on the five specific features provided during training.

#Executive Summary
**Average Prediction Error:** The model operates with an average deviation of 0.0014 TL.
Operational Use: The system is suitable for providing preliminary price estimations in the second-hand vehicle market.

**Scalability:** Performance and reliability can be further enhanced by incorporating larger datasets and additional features like "Damage History" or "Fuel Type."